misc/migrationAides/backupMviDb.py

```diff
@@ -0,0 +1,388 @@
+        cmdArgs = ["oc", "login", args.cluster]
+        if args.ocpToken is not None:
+            cmdArgs.extend(["--token", args.ocpToken])
+        else:
+            cmdArgs.extend(["--username", args.ocpUser, "--password", args.ocpPasswd])
+        logging.info(f"logging into cluster '{args.cluster}'")
+
+        process = subprocess.run(cmdArgs, capture_output=True)
+        if process.returncode != 0:
+            logging.error(f"Failed to login to cluster {cmdArgs}")
+            logging.error(f"output = {process.stderr}")
+            raise OcpException(f"Failed to login to cluster {args.cluster}.")
+
+        setOcpProject()
+
+
+def setOcpProject():
+    cmdArgs = ["oc", "project", args.project]
+    logging.debug(f"Setting project to '{args.project}'")
+
+    process = subprocess.run(cmdArgs, capture_output=True)
+    if process.returncode != 0:
+        logging.error(f"Failed to login to cluster {cmdArgs}")
+        logging.error(f"output = {process.stderr}")
+        raise OcpException(f"Failed to connect to project {args.project}")
+
+
+def connectToMongo():
+    logging.info("Connecting to mongo")
+
+    user, passwd = getMongoDbCredentials()
+    tunnelToMongo()
+    loginToDb(user, passwd)
+
+
+def getMongoDbCredentials():
+    """ Returns mongoDB username and password based upon provided access info."""
+
+    # If mongoDb credentials provided to the script, use those
+    if mongoDbCreds is not None:
+        return mongoDbCreds["userName"], mongoDbCreds["password"]
+
+    cmdArgs = ["oc", "get", "secret", "vision-secrets", "-o", "json"]
+
+    # otherwise get credentials from the cluster
+    process = subprocess.run(cmdArgs, capture_output=True)
+    if process.returncode == 0:
+        jsonData = json.loads(process.stdout)
+        userName = base64.b64decode(jsonData["data"]["mongodb-admin-username"]).decode("utf-8")
+        password = base64.b64decode(jsonData["data"]["mongodb-admin-password"]).decode("utf-8")
+        logging.debug(f"user={userName}, pw={password}")
+    else:
+        logging.error(f"Failed to get Mongo info -- {cmdArgs}")
+        logging.error(f"output = {process.stderr}")
+        raise OcpException(f"Could not get Mongo connection info.")
+
+    return userName, password
+
+
+def tunnelToMongo():
+    """ Sets up a tunnel the mongoDB if backing up a post 8.0.0 database."""
+    if mongoDbCreds:
+        return
+
+    pod = getMongoPod()
+    if pod is None:
+        raise OcpException("Could not find MongoDB Pod.")
+    establishTunnel(pod, 27017, 27017)
+
+
+def getMongoPod():
+    """ Returns the pod name of the running mongoDB pod in an OCP cluster."""
+    cmdArgs = ["oc", "get", "pods"]
+    pod = None
+
+    process = subprocess.Popen(cmdArgs, stdout=subprocess.PIPE)
+    for line in process.stdout:
+        string = line.decode();
+        logging.debug(string)
+        if re.search("-mongodb-", string):
+            pod = string.split(" ")[0]
+            break
+    process.wait()
```

Comment:
just wondering why a pipe for the subprocess is done here and not for the other commands run above like for the mongo password.

Reply:
PIPE does 2 things. 
 1. It leaves the process running when you return from the call, where as if you just capture STDOUT, the call will not return until the process completes and all of STDOUT is available.  The port-forward command stays running as it handles the forwarding.
 2. It allows you to read STDOUT a line at a time. In this case, I probably could have used capture   STDOUT and just looped through stdout to find the mongo pod. But, my thought process at the time was to read the output a line at a time and find the one that I am interested in as it "pops out".